go/types/value_decoder.go

```diff
@@ -49,23 +50,26 @@ func (r *valueDecoder) readType() *Type {
 	return t
 }
 
-func (r *valueDecoder) readTypeInner() *Type {
+func (r *valueDecoder) readTypeInner(seenStructs map[string]*Type) *Type {
 	k := r.readKind()
 	switch k {
 	case ListKind:
-		return makeCompoundType(ListKind, r.readType())
+		return makeCompoundType(ListKind, r.readType(seenStructs))
 	case MapKind:
-		return makeCompoundType(MapKind, r.readType(), r.readType())
+		return makeCompoundType(MapKind, r.readType(seenStructs), r.readType(seenStructs))
 	case RefKind:
-		return makeCompoundType(RefKind, r.readType())
+		return makeCompoundType(RefKind, r.readType(seenStructs))
 	case SetKind:
-		return makeCompoundType(SetKind, r.readType())
+		return makeCompoundType(SetKind, r.readType(seenStructs))
 	case StructKind:
-		return r.readStructType()
+		return r.readStructType(seenStructs)
 	case UnionKind:
-		return r.readUnionType()
+		return r.readUnionType(seenStructs)
 	case CycleKind:
-		return MakeCycleType(r.readUint32())
+		name := r.readString()
```

Comment:
This changes `Cycle<0>` to `Cycle<Name>`. I'll have to update the code in a few more places to ensure consistency.